[Bug]: Build fails on debian12

### 🐞 Bug

Building from source on debian12 results in a static assertion failure.

maybe related to
https://github.com/stack-of-tasks/pinocchio/commit/dfd09d691f61089696dfebdf660ea5e5135afa40
https://github.com/stack-of-tasks/pinocchio/commit/d6dd5b83971118544042ebec0c4db87d81867548

### Reproduction steps

```bash
colcon build \
      --cmake-args --no-warn-unused-cli -DBUILD_EXAMPLES=OFF -DBUILD_PYTHON_INTERFACE=OFF \
        -DBUILD_WITH_COLLISION_SUPPORT=OFF \
      --packages-select pinocchio
```

more specifically, from 
https://github.com/ros-controls/ros2_control_ci/actions/runs/25821777325/job/75864643148?pr=787

``` bash
git clone https://github.com/ros-controls/ros2_control_ci.git -b fix-debian-coal
docker buildx build --build-arg ROS_DISTRO=lyrical --build-arg EXCLUDE_PACKAGES=gz_ros2_control gz_ros2_control_demos ign_ros2_control mujoco_ros2_control ros_gz_sim ros_gz_bridge rviz2 eigenpy hpp-fcl coal --file ros2_debian/Dockerfile.debian12 .
```

I can try to reproduce it with a minimum dockerfile if that would help.

### Relevant log output / Error message

```shell
#18 2477.4 + colcon build --cmake-args --no-warn-unused-cli -DBUILD_EXAMPLES=OFF -DBUILD_PYTHON_INTERFACE=OFF -DBUILD_WITH_COLLISION_SUPPORT=OFF --continue-on-error --packages-select pinocchio
#18 2477.4 --- stderr: pinocchio
#18 2477.4 In file included from /opt/ros2_ws/src/generated/pinocchio/include/pinocchio/constraints.hpp:107,
#18 2477.4                  from /opt/ros2_ws/src/generated/pinocchio/include/pinocchio/algorithm/impulse-dynamics.hpp:24,
#18 2477.4                  from /opt/ros2_ws/src/generated/pinocchio/src/algorithm/impulse-dynamics.cpp:8:
#18 2477.4 /opt/ros2_ws/src/generated/pinocchio/include/pinocchio/src/constraints/utils.hxx: In static member function ‘static void pinocchio::internal::ComputeBlockDiagonalPatternImpl<ConstraintModel, <template-parameter-1-2> >::run(const ConstraintModel&, BlockInfoVector&, pinocchio::BlockDiagonalDispatcherTag<op>)’:
#18 2477.4 /opt/ros2_ws/src/generated/pinocchio/include/pinocchio/src/constraints/utils.hxx:950:11: error: static assertion failed: ComputeBlockDiagonalPatternImpl not implemented for this constraint.
#18 2477.4   950 |           false, "ComputeBlockDiagonalPatternImpl not implemented for this constraint.");
#18 2477.4       |           ^~~~~
#18 2477.4 gmake[2]: *** [src/CMakeFiles/pinocchio_default.dir/build.make:149: src/CMakeFiles/pinocchio_default.dir/algorithm/impulse-dynamics.cpp.o] Error 1
#18 2477.4 gmake[2]: *** Waiting for unfinished jobs....
#18 2477.4 gmake[1]: *** [CMakeFiles/Makefile2:2068: src/CMakeFiles/pinocchio_default.dir/all] Error 2
#18 2477.4 gmake: *** [Makefile:149: all] Error 2
#18 2477.4 ---
```

### System Info

debian12, ROS lyrical from source with rosinstall_generator

### Checklist

- [x] I have checked that all my packages are installed from the same package manager (conda-forge, Nix, PyPI, robotpkg, ROS, ...)
- [x] I have checked that there is no similar [issue](https://github.com/stack-of-tasks/pinocchio/issues)/[discussion](https://github.com/stack-of-tasks/pinocchio/discussions) in the repo
- [x] I have provided a minimal and working example to reproduce the bug
- [x] I have used [Markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces/compiler errors.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,7 +10,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
 
-- nix: switch to flakoboros
+### Fixed
+
+- Fix build issue with g++ 12 ([#2890](https://github.com/stack-of-tasks/pinocchio/pull/2890))
+
+### Changed
+
+- nix: switch to flakoboros ([#2882](https://github.com/stack-of-tasks/pinocchio/pull/2882))
 
 ## [4.0.0] - 2026-04-13
 

--- a/include/pinocchio/src/constraints/utils.hxx
+++ b/include/pinocchio/src/constraints/utils.hxx
@@ -943,12 +943,12 @@ namespace pinocchio
         PINOCCHIO_UNUSED_VARIABLE(dispatcher);
 
         // For some reason this assert is always evaluated when building
-        // with g++ 11.
-        // TODO Remove when Ubuntu 22.04 is no more supported.
-#if !defined(__GNUC__) || __GNUC__ >= 12
+        // with g++ 11 and 12.
+        // TODO Remove when Ubuntu 22.04 and Debian 12 (ROS) is no more supported.
+#if !defined(__GNUC__) || __GNUC__ >= 13
         static_assert(
           false, "ComputeBlockDiagonalPatternImpl not implemented for this constraint.");
-#endif // !defined(__GNUC__) || __GNUC__ >= 12
+#endif // !defined(__GNUC__) || __GNUC__ >= 13
       }
     };
 

